x-pack/plugins/security/server/routes/authorization/roles/post.ts

```diff
@@ -0,0 +1,127 @@
+  noop?: string[];
+  created?: string[];
+  updated?: string[];
+  errors?: {
+    count: number;
+    details: RolesErrorsDetails;
+  };
+}
+
+export function definePostRolesRoutes({
+  router,
+  authz,
+  getFeatures,
+  getFeatureUsageService,
+}: RouteDefinitionParams) {
+  router.post(
+    {
+      path: '/api/security/roles',
+      options: {
+        summary: 'Create or update roles',
+      },
+      validate: {
+        body: getPostPayloadSchema(() => {
+          const privileges = authz.privileges.get();
+          return {
+            global: Object.keys(privileges.global),
+            space: Object.keys(privileges.space),
+          };
+        }),
+      },
+    },
+    createLicensedRouteHandler(async (context, request, response) => {
+      try {
+        const esClient = (await context.core).elasticsearch.client;
+        const features = await getFeatures();
+
+        const { roles } = request.body;
+        const validatedRolesNames = [];
+        const kibanaErrors: RolesErrorsDetails = {};
+
+        for (const [roleName, role] of Object.entries(roles)) {
+          const { validationErrors } = validateKibanaPrivileges(features, role.kibana);
+
+          if (validationErrors.length) {
+            kibanaErrors[roleName] = {
+              type: 'kibana_privilege_validation_exception',
+              reason: `Role cannot be updated due to validation errors: ${JSON.stringify(
+                validationErrors
+              )}`,
+            };
+
+            continue;
+          }
+
+          validatedRolesNames.push(roleName);
+        }
+
+        const rawRoles = await esClient.asCurrentUser.security.getRole(
+          { name: validatedRolesNames.join(',') },
+          { ignore: [404] }
+        );
+
+        const esRolesPayload = Object.fromEntries(
+          validatedRolesNames.map((roleName) => [
+            roleName,
+            transformPutPayloadToElasticsearchRole(
+              roles[roleName],
+              authz.applicationName,
+              rawRoles[roleName] ? rawRoles[roleName].applications : []
+            ),
+          ])
+        );
+
+        const esResponse = await esClient.asCurrentUser.transport.request<ESRolesResponse>({
+          method: 'POST',
+          path: '/_security/role',
+          body: { roles: esRolesPayload },
+        });
+
+        for (const roleName of [...(esResponse.created ?? []), ...(esResponse.updated ?? [])]) {
+          if (roleGrantsSubFeaturePrivileges(features, roles[roleName])) {
+            getFeatureUsageService().recordSubFeaturePrivilegeUsage();
+          }
+        }
+
+        const { created, noop, updated, errors: esErrors } = esResponse;
+        const hasAnyErrors = Object.keys(kibanaErrors).length || esErrors?.count;
+
+        return response.ok({
+          body: {
+            created,
+            noop,
+            updated,
+            ...(hasAnyErrors && {
+              errors: { ...kibanaErrors, ...(esErrors?.details ?? {}) },
```

Comment:
__observation:__ the `reason` for Kibana specific errors formatted in a weird way in our dev tools (too many leading/trailing quotes). I suppose it's just how Kibana Dev tools app renders strings with nested quotes and not something we need to worry about?

![image](https://github.com/user-attachments/assets/16a6e576-4157-4a7d-b52a-617f4286d690)

Reply:
True, it is the same with triple quotes for out `PUT` endpoint

Reply:
Got it, thanks for confirming.